File: 678-valid-parenthesis-string.cs
public class Solution {
    public bool CheckValidString(string s) {
                var leftStack = new Stack<int>();
                var starStack = new Stack<int>();
                for (int i = 0; i < s.Length; i++) {
                    if (s[i] == '(') {
                        leftStack.Push(i);
                    } else if (s[i] == '*') {
                        starStack.Push(i);
                    } else {
                        if (leftStack.Count != 0) {
                            leftStack.Pop();
                        } else if (starStack.Count != 0) {
                            starStack.Pop();
                        } else {
                            return false;
                        }
                    }
                }
                if (leftStack.Count > starStack.Count) return false;
                while (leftStack.Count != 0 & starStack.Count != 0) {
                    if (leftStack.Peek() > starStack.Peek())
                        return false;
                    else {
                        leftStack.Pop();
                        starStack.Pop();
                    }
                }
                if (leftStack.Count == 0) return true;
                return false;
    }
}

File: .vscode/678-valid-parenthesis-string.go
package main

func checkValidString(s string) bool {
	var leftStack, starStack []int
	for i, ch := range s {
		if ch == '(' {
			leftStack = append(leftStack, i)
		} else if ch == '*' {
			starStack = append(starStack, i)
		} else {
			if len(leftStack) != 0 {
				leftStack = leftStack[:len(leftStack)-1]
			} else if len(starStack) > 0 {
				starStack = starStack[:len(starStack)-1]
			} else {
				return false
			}

		}
	}
	i := len(leftStack) - 1
	for j := len(starStack) - 1; i >= 0 && j >= 0; i, j = i-1, j-1 {
		if leftStack[i] > starStack[j] {
			return false
		}
	}
	return i < 0
}
